infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java

```diff
@@ -600,8 +612,35 @@ public void accept(ProbeResult probeResult) {
     }
   }
 
+  /** Listens container's events and terminates workspace if unrecoverable event occurs. */
+  public class UnrecoverableEventHanler implements ContainerEventHandler {
+
+    @Override
+    public void handle(ContainerEvent event) {
+      if (isUnrecoverable(event)) {
+        String reason = event.getReason();
+        String message = event.getMessage();
+        String podName = event.getPodName();
+        try {
+          internalStop(emptyMap());
```

Comment:
I think it doesn't work correctly until k8s runtime start interruption will be implemented. 
I guess client will receive two workspace status events:
- one that workspace failed to start because of Unrecoverable;
- a second one that start timeout exceeded.

Let me check that. BTW I think we can merge these changes with the described issue and it will be fixed with implemented interruption of k8s runtime start.

Reply:
I've checked. So here is the Che Server logs
```

  | 2018-04-17 08:02:45,912[io-8080-exec-10]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 236]   - Starting workspace 'che/qwe' with id 'workspacea75oxtgryzdjq5iq' by user 'che'
  /////////////////// Here Client received that workspace failed to start but Che Server doesn't contain the corresponding message ///////////////////////////
  /////////////////// ~3 minutes later reached workspace start timeout ///////////////////////////
  | 2018-04-17 08:05:18,006[aceSharedPool-0]  [WARN ] [.i.k.KubernetesInternalRuntime 177]  - Failed to start Kubernetes runtime of workspace workspacelhu15ifhded6jrxf. Cause: Server 'terminal' in machine 'dev-machine' not available.
  | 2018-04-17 08:05:18,006[ServersChecker]   [ERROR] [.i.k.KubernetesInternalRuntime 335]  - Unable to update status of the machine '%s:%s'. Cause: %s
  | 2018-04-17 08:05:18,066[aceSharedPool-0]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 292]   - Workspace 'che:wksp-g9f7' with id 'workspacelhu15ifhded6jrxf' start failed
  | 2018-04-17 08:05:18,070[aceSharedPool-0]  [WARN ] [o.e.c.a.w.s.WorkspaceManager 423]    - Cannot set error status of the workspace workspacelhu15ifhded6jrxf. Error is: Workspace with id 'workspacelhu15ifhded6jrxf' doesn't exist
```